frontend/model/contracts/chatroom.js

```diff
@@ -530,19 +550,28 @@ sbp('chelonia/defineContract', {
         manifestCid: string,
         messageSender: string
       })),
-      process ({ data, innerSigningContractID }, { state }) {
-        const msgIndex = findMessageIdx(data.hash, state.messages)
-        if (msgIndex >= 0) {
-          const oldAttachments = state.messages[msgIndex].attachments
+      process ({ data }, { state }) {
+        const { hash, manifestCid } = data
+        const fnDeleteAttachment = (message) => {
+          const oldAttachments = message.attachments
           if (Array.isArray(oldAttachments)) {
             const newAttachments = oldAttachments.filter(attachment => {
-              return attachment.downloadData.manifestCid !== data.manifestCid
+              return attachment.downloadData.manifestCid !== manifestCid
             })
-            Vue.set(state.messages[msgIndex], 'attachments', newAttachments)
+            Vue.set(message, 'attachments', newAttachments)
           }
         }
+
+        const msgIndex = findMessageIdx(hash, state.messages)
+        if (msgIndex >= 0) {
+          fnDeleteAttachment(state.messages[msgIndex])
+        }
+        const pinnedMsgIndex = findMessageIdx(hash, state.pinnedMessages)
+        if (pinnedMsgIndex >= 0) {
+          fnDeleteAttachment(state.pinnedMessages[pinnedMsgIndex])
+        }
```

Comment:
Hmm, so the code is duplicating the pinned messages completely, even the attachments?

Is it possible to rewrite the code so that the pinned messages do not duplicate the data, but instead only reference the messages in `state.messages` by hash instead? This way we can DRY the code, save on memory,  and not worry about updating data in two different places.

As it is right now, the code has to have all of these changes to add these `fn*` functions all over the place to perform the same operation twice. Can we please instead add the message hashes to the pinned messages list?

Reply:
The reason why I made `fn*` function and called it twice for `state.messages` and `state.pinnedMessages` is that we don't save all messages in `state.messages`.
When users open channel, they should see `N pinned` in the chat page and also by clicking it, they should see all the pinned messages.

We can also save only message hashes. If so, it's fine to display `N pinned`, but when user clicks it, he needs to render all the events from the first pinned message to the latest event. And I think this is bad in terms of performance. That's why I saved the all pinned messages, instead of their hashes.

Reply:
I see, ok that make sense.

In that case, since the code is so similar, do you think we can DRY it a bit further? E.g. something like this maybe?

```js
        [state.messages, state.pinnedMessages].forEach(messageArray => {
          const msgIndex = findMessageIdx(hash, messageArray)
          if (msgIndex >= 0) {
            fnDeleteAttachment(messageArray[msgIndex])
          }
        })
```

BTW, that was suggested by an LLM. I used [Cursor](https://www.cursor.com/) and highlighted the code, and asked it to DRY it so that `fnDeleteAttachment` was only called once. 😄